docs/RFCS/data_structures_for_logical_planning.md

```diff
@@ -0,0 +1,1001 @@
+the introduction of expression classes, so that expressions do not
+form a "textbook" tree in memory any more.
+
+The scope of this RFC is restricted to:
+
+- give an overall idea to the human reader of how a memo works and
+  motivate its choice, so as to enable more team members to
+  participate in the exploratory effort.
+- present a path to integrate a memo structure into the current
+  CockroachDB code base.
+
+In particular, this RFC does not intend to detail the particular
+algorithms that will be used to populate the memo and search for
+optimal queries, nor does it intend to fully specify the set of
+logical and physical properties of a class of expressions.
+
+# Context of this document
+
+This RFC is based on the following prior work:
+
+- [Tech note: Guiding principles for the SQL middle-end and back-end in CockroachDB](../tech-nodes/sql-principles.md)
+- [Tech note / meta-RFC: Upcoming SQL changes / Evolving the SQL layer in CockroachDB](https://github.com/cockroachdb/cockroach/pull/18977)
+- [RFC: SQL Logical planning](https://github.com/cockroachdb/cockroach/pull/19135) in particular the terminology
+  set forth there, for example the "prep", "rewrite" and "search" phases.
+- [RFC: Algebraic Data Types in Go](20170517_algebraic_data_types.md) because we're going to reuse the mechanisms there.
+
+The readers are encouraged to read these prior work RFCs, in
+particular the RFC on logical planning by Peter, and make their own
+experiments in code to accompany their understanding of the concepts
+presented in this RFC.
+
+As of this writing, the following code experiments can be used to
+accompany the reading in this RFC:
+
+- bare-bones, introductory code with many comments by knz: http://github.com/knz/sqlexp
+- a more advanced prototype in Go by Peter: http://github.com/petermattis/opttoy
+
+In the plan set forward with "upcoming SQL changes" the present RFC
+corresponds to the items:
+
+- SQL leaf data
+- Introduce scoping and free variables in name resolution
+- Apply relational operator
+- Plan equivalency classes
+- Make plans stateless / context-free
+
+i.e. acting on this RFC will move forward on all these 5 items at
+once, and complete several of them.
+
+Finally, this RFC is not *based* on the [IR
+RFC](https://github.com/cockroachdb/cockroach/pull/10055) but will
+instead serve as new *prelude* to it. The text in the IR RFC that
+pertains to this one will be migrated here, and the remaining text
+will be later updated to reflect the choices made here.
+
+# Motivation
+
+This RFC is motivated by:
+
+1) a need for new data structures to reduce various overheads of the
+   current technology.
+2) a desire to align the architecture of CockroachDB with the state-of-the-art,
+   which will ease the adoption of algorithms and ideas from the literature.
+3) a need to set up the scene for multiple other RFCs to come.
+4) a need to aid the reading of various prototypes
+   developed in the team to experiment with these concepts.
```

Comment:
This is what I was looking for, but "reduce various overheads" is not compelling enough. Thanks